src/lib.rs

```diff
@@ -192,59 +312,136 @@ impl<'object, Endian> DebugInfo<'object, Endian>
                 }
             }
         }
-        None
+
+        loc.map(move |(f, l)| (f, l, func.map(|u| u.0.name.to_string_lossy())))
     }
 }
 
-// TODO: most of this should be moved to the main library.
-struct Unit<'input> {
-    address_size: u8,
-    ranges: Vec<gimli::Range>,
-    line_offset: gimli::DebugLineOffset,
-    comp_dir: Option<&'input std::ffi::CStr>,
-    comp_name: Option<&'input std::ffi::CStr>,
+enum Target<'input> {
+    Line(LineUnit<'input>),
+    Function(FunctionUnit<'input>),
```

Comment:
The `Line` variant is larger than the `Function` variant (88 bytes vs 56 bytes), but there'll be many more functions than compilation units, so there'll be some memory saving. Also, the usage in `locate()` seems to treat them separately, and the flow control is a bit different (you want the most specific function, but the first line match is ok), so I wasn't sure there was much benefit to having them in the same enum. The only common bit I can see is parsing the range attributes, but that could be moved to a helper function that they both call. Not something I feel strongly about though.